src/ucp/proto/proto.h

```diff
@@ -0,0 +1,184 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2020.  ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#ifndef UCP_PROTO_H_
+#define UCP_PROTO_H_
+
+#include "lane_type.h"
+
+#include <ucp/core/ucp_types.h>
+#include <ucs/datastruct/linear_func.h>
+#include <ucs/datastruct/string_buffer.h>
+
+
+/* Maximal number of lanes per protocol */
+#define UCP_PROTO_MAX_LANES         UCP_MAX_LANES
+
+
+/* Maximal number protocol performance ranges */
+#define UCP_PROTO_MAX_PERF_RANGES   (UCP_PROTO_MAX_LANES + 1)
+
+
+/* Maximal size of protocol private data */
+#define UCP_PROTO_PRIV_MAX          1024
+
+
+/* Maximal number protocols in total */
+#define UCP_PROTO_MAX_COUNT         64
+
+
+/* Special value for non-existent protocol */
+#define UCP_PROTO_ID_INVALID        ((ucp_proto_id_t)-1)
+
+
+/* Protocol identifier */
+typedef unsigned ucp_proto_id_t;
+
+
+/* Bitmap of protocols */
+typedef uint64_t ucp_proto_id_mask_t;
+
+
+/**
+ * Key for looking up protocol configuration by operation parameters
+ */
+typedef struct {
+    uint8_t                 op_id;      /* Operation ID */
+    uint8_t                 op_flags;   /* Operation flags */
+    uint8_t                 dt_class;   /* Datatype */
+    uint8_t                 mem_type;   /* Memory type */
+    uint8_t                 sys_dev;    /* System device */
+    uint8_t                 sg_count;   /* Number of non-contig scatter/gather
+                                           entries. If the actual number is larger
+                                           than UINT8_MAX, UINT8_MAX is used. */
+    uint8_t                 padding[2]; /* Make structure size be sizeof(uint64_t) */
+} UCS_S_PACKED ucp_proto_select_param_t;
+
+
+/**
+ * Protocol and its private configuration
+ */
+typedef struct {
+    const ucp_proto_t       *proto;  /* Protocol definition */
+    const void              *priv;   /* Protocol private configuration space */
+} ucp_proto_config_t;
+
+
+/*
+ * Performance estimation for a range of message sizes
+ */
+typedef struct {
+    size_t                  max_length; /* Maximal message size */
```

Comment:
why is it needed here?
should not it be a range for defining piecewise linear funcs?

Reply:
this is it: ucp_proto_perf_range_t is a single "piece". max_length is the upper bound (inclusive) of the piece. "perf" is the linear function which represents the performance for this piece.
the first piece (in ucp_proto_caps_t) starts from "min_length".

Reply:
and the second piece starts from the `max_length` of the first?

Reply:
piece N+1 starts from piece[N].max_length + 1